Alter way to catch missing columns error #469

FlorianK13 opened the pull request:
Instead of getting the column name from the
error message, the missing columns are now received by comparing the existing database table with the new dataframe.


### PR-Assignee
- [x] 🐙 Follow the workflow in [CONTRIBUTING.md](https://github.com/OpenEnergyPlatform/open-MaStR/blob/production/CONTRIBUTING.md)
- [x] 📝 Update the [CHANGELOG.md](https://github.com/OpenEnergyPlatform/open-MaStR/blob/production/CHANGELOG.md)
- [x] 📙 Update the documentation

### Reviewer
- [ ] 🐙 Follow the [Reviewer Guidelines](https://github.com/OpenEnergyPlatform/open-MaStR/blob/production/CONTRIBUTING.md#40-let-someone-else-review-your-pr)
- [ ] 🐙 Provided feedback and show sufficient appreciation for the work done

FlorianK13 committed:
Alter way to catch missing columns error #469

Instead of getting the column name from the
error message, the missing columns are now received by comparing
the existing database table with the new dataframe.

FlorianK13 committed:
Update CHANGELOG.md

k-schmid commented on open_mastr/xml_download/utils_write_to_database.py:
```diff
@@ -302,44 +302,51 @@ def write_single_entries_until_not_unique_comes_up(
 
 
 def add_missing_column_to_table(
```
I would suggest to rename it to add_missing_columns_to_table

FlorianK13 replied:
I agree and changed it.

k-schmid commented on open_mastr/xml_download/utils_write_to_database.py:
```diff
@@ -302,44 +302,51 @@ def write_single_entries_until_not_unique_comes_up(
+    err: Error,
+    engine: sqlalchemy.engine.Engine,
+    xml_tablename: str,
+    df: pd.DataFrame,
```
Why not passing required_columns: List[str] instead of the whole dataframe?

FlorianK13 replied:
I think that I prefer to have this in the sub function, so that the code within the main function stays more abstract. What would be your argument for creating the List in the main function?

k-schmid replied:
My main argument is cleanness: The objective of the function is to add missing columns to the table. Therefor a list of required columns should ideally be one of the inputs (and would be what I would assume).

Second but weaker argument is safeness: There could be potential side effects on the df if one would manipulate it in the method (as it is a mutable object). Thus it is safer to just pass what is needed.

FlorianK13 replied:
Adapted in latest commit

k-schmid commented on open_mastr/xml_download/utils_write_to_database.py:
```diff
@@ -302,44 +302,51 @@ def write_single_entries_until_not_unique_comes_up(
 
 def add_missing_column_to_table(
-    err: Error, engine: sqlalchemy.engine.Engine, xml_tablename: str
+    err: Error,
```
Is the err still needed?

FlorianK13 replied:
Nope, changed it in the last commit.

k-schmid commented on open_mastr/xml_download/utils_write_to_database.py:
```diff
@@ -212,18 +211,18 @@ def add_table_to_database(
                     continueloop = False
         except sqlalchemy.exc.OperationalError as err:
-            add_missing_column_to_table(err, engine, xml_tablename)
+            add_missing_column_to_table(err, engine, xml_tablename, df)
```
In my opinion it would make more sense to move the add_missing_column_to_table up to do it before writing data to the table.
Something like:
```
        add_missing_column_to_table(engine, xml_tablename, required_columns)
        try:
            with engine.connect() as con:
                with con.begin():
                    df.to_sql(
                        sql_tablename,
                        con=con,
                        index=False,
                        if_exists=if_exists,
                        dtype=dtypes_for_writing_sql,
                    )
        except sqlalchemy.exc.IntegrityError:
            # error resulting from Unique constraint failed
            df = write_single_entries_until_not_unique_comes_up(
                df=df, xml_tablename=xml_tablename, engine=engine
               
```

FlorianK13 committed:
Delete while True loop #469

Better have a for loop with a large number which will still stop
at some point.

FlorianK13 committed:
Move add_missing_columns outside of loop #469

Additionally rename function and change
while True loop to for loop

FlorianK13 committed:
Merge branch 'bugfix-469-add_missing_column_to_table' of github.com:OpenEnergyPlatform/open-MaStR into bugfix-469-add_missing_column_to_table

FlorianK13 commented:
Tests are failing due to #466 I assume.

FlorianK13 committed:
Pass columns list instead of df #469

FlorianK13 committed:
Merge branch 'bugfix-466-web-service-update' into bugfix-469-add_missing_column_to_table

FlorianK13 commented:
Do you think we can merge @chrwm ?

FlorianK13 committed:
Merge branch 'develop' into bugfix-469-add_missing_column_to_table

chrwm reviewed (approved):
Looks good!
